kube-runtime/src/scheduler.rs

```diff
@@ -444,7 +473,72 @@ mod tests {
                 },
             ])
             .on_complete(sleep(Duration::from_secs(5))),
+            None,
         );
         assert_eq!(scheduler.map(|msg| msg.0).collect::<Vec<_>>().await, vec![1]);
     }
+
+    #[tokio::test]
+    async fn scheduler_should_discard_same_message_within_debounce_period() {
+        pause();
+
+        let now = Instant::now();
+        let (mut sched_tx, sched_rx) = mpsc::unbounded::<ScheduleRequest<SingletonMessage>>();
+        let mut scheduler = scheduler(sched_rx, Some(Duration::from_secs(5)));
+
+        sched_tx
+            .send(ScheduleRequest {
+                message: SingletonMessage(1),
+                run_at: now + Duration::from_secs(1),
+            })
+            .await
+            .unwrap();
+        assert!(poll!(scheduler.next()).is_pending());
+        advance(Duration::from_secs(2)).await;
+        assert_eq!(scheduler.next().now_or_never().unwrap().unwrap().0, 1);
+        // make sure that we insert the popped message into expired
+        assert_eq!(scheduler.expired.len(), 1);
```

Comment:
This is not quite how i envisioned debouncing would work. This basically takes the first and eliminates the last which is functionally wrong (the user should always see the last request as it is the most up to date).

Normally debounce works by adding an additional `5s` (say) to the initial schedule, and in that period it does not run the first reconcile. If multiple requests happen within that 5s period, the 5s wait time gets reset so that we only actually run the reconciler 5s after the last uninterrupted request.

Reply:
so is it more like if a request came in at time `T` asking the scheduler to emit it at `T+10` and another request came in at time `T+1` asking to be emitted at `T+11`, then we need to consider the latter request, provided that no other requests came in until `T+6`? so basically its not `run_at` that's to be considered here but the time at which the request arrives to the scheduler?

Reply:
The requested run time isn't disregarded. It's more that we are giving a passive leeway in both the vacant and occupied case, so that the `scheduler` does not emit immediately no matter what.

We already deduplicate in the `T+10` and `T+11` case you wrote because they are far enough in the future that neither have run (and we pick the last). But we don't get the chance to deduplicate when people schedule for "now".

I think we can, in the `Vacant` case, add the debounce time to the scheduled run time and rely on the natural deduplication to do its job. But  we must account for the shift in the `Occupied` case to ensure we actually pick the last still (and also add debounce time there when resetting).

Reply:
> Normally debounce works by adding an additional 5s (say) to the initial schedule, and in that period it does not run the first reconcile. If multiple requests happen within that 5s period, the 5s wait time gets reset so that we only actually run the reconciler 5s after the last uninterrupted request.

That's dangerous, it means that someone writing more frequently than the debounce period can lock you out of reconciling completely...

Reply:
> That's dangerous, it means that someone writing more frequently than the debounce period can lock you out of reconciling completely...

yep. that's why the period should be suggested as a few seconds at most. generally in kubernetes objects go through phases quickly when they are setting up, but then stabilise - and that's when you get your event (if you've chosen to debounce).

in the periodic object updater case, say something like a `Node` controller (which updates its status every like 30s by design), then it will lock you out if you put the debounce period greater than that. but also, if you put the number that high, you will have a uselessly slow reconciler even when developing.

i think we need to emphasize that debounce is a tool, and it is not meant to prevent all types of reconciliations retriggers (use `predicates::generation` to prevent status changes), and it is potentially a user error to put a too high number here.

the alternative is to maybe track the amount of delay we have caused and not add the debounce time if we've already added like 10x the debounce time? not sure how viable that is.

Reply:
The goal is to run the right amount of reconciles though. If you know your constraints you won't under-run.

We can keep it off by default if you prefer. But I would really like to at least have the option to set this because I have hot controllers running thousand reconciliations per minute, and improvements towards cutting down the noise is helpful.